lib/parsers/yarn-lock-parse-base.ts

```diff
@@ -0,0 +1,165 @@
+import * as _ from 'lodash';
+import * as pMap from 'p-map';
+
+import {
+    LockfileParser, PkgTree, DepTreeDep, Dep, ManifestFile,
+    getTopLevelDeps, Lockfile, LockfileType, createDepTreeDepFromDep,
+} from './';
+import {setImmediatePromise} from '../set-immediate-promise';
+import {
+    InvalidUserInputError,
+    OutOfSyncError,
+} from '../errors';
+
+const EVENT_PROCESSING_CONCURRENCY = 5;
+
+export type YarnLockFileTypes = LockfileType.yarn | LockfileType.yarn2;
+
+export interface YarnLockDeps {
+    [depName: string]: YarnLockDep;
+}
+
+export interface YarnLockBase<T extends YarnLockFileTypes> {
+    type: string;
+    object: YarnLockDeps;
+    dependencies?: YarnLockDeps;
+    lockfileType: T;
+}
+
+export interface YarnLockDep {
+    version: string;
+    dependencies?: {
+        [depName: string]: string;
+    };
+    optionalDependencies?: {
+        [depName: string]: string;
+    };
+}
+
+export abstract class YarnLockParseBase<T extends YarnLockFileTypes> implements LockfileParser {
+
+    private treeSize: number;
+    private eventLoopSpinRate = 20;
+
+    constructor(private type: T) {
+        // Number of dependencies including root one.
+        this.treeSize = 1;
+    }
+
+    public abstract parseLockFile(lockFileContents: string): Lockfile;
+
+    public async getDependencyTree(
+        manifestFile: ManifestFile, lockfile: Lockfile, includeDev = false,
+        strict = true ): Promise<PkgTree> {
+        if (lockfile.type !== this.type) {
+            throw new InvalidUserInputError('Unsupported lockfile provided. ' +
+                'Please provide `package-lock.json`.');
```

Comment:
```suggestion
                'Please provide a `yarn.lock` file.');
```

Reply:
I will leave that decision to the maintainers